### ISSUE TITLE
`require` statement does not emit the given error-message

### ISSUE THREAD

**Issue by barakman** [trufflesuite/truffle#4900]
## Issue

I believe that this is a bug in the compiler, in which, under very specific circumstances, a `require` statement does not emit the given error-message.

So I have initially posted it on [Solc GitHub](https://github.com/ethereum/solidity/issues/12819).

However, the issue might actually be on Truffle and/or Ganache side, so I am posting it here as well.

## Environment

- Compiler version: 0.8.12
- Ganache version: 7.0.1
- Truffle version: 5.4.33
- Operating system: MacOS
- node version (`node --version`): v16.13.0
- npm version (`npm --version`): 8.1.0
- yarn version (`yarn --version`): 1.22.10

## Steps to Reproduce

```solidity
contract MyContract {
    struct Set {
        address[] _values;
        mapping(address => uint256) _indexes;
    }

    Set private addressSet;

    function test(address[] calldata addresses) external {
        for (uint256 i = 0; i < addresses.length; i++)
            require(add(addresses[i]), "already exists");
    }

    function add(address value) internal returns (bool) {
        return _add(addressSet, value);
    }

    function _add(Set storage set, address value) private returns (bool) {
        if (set._indexes[value] == 0) {
            set._values.push(value);
            set._indexes[value] = set._values.length;
            return true;
        } else {
            return false;
        }
    }
}
```
When function `test(addresses)` is executed with an address which has already been passed to it in a previous execution, it should revert with an error-message `"already exists"`.

Indeed, this error-message is emitted when we do so in two different transactions:
```
await myContract.test([ADDRESS1, ADDRESS2]);
await myContract.test([ADDRESS1, ADDRESS2]); // reverts with `already exists`
```
However, when we do so in a single transaction:
```
await myContract.test([ADDRESS1, ADDRESS1]);
```
The transaction reverts with `exited with an error (status 0) after consuming all gas`.
I suppose that this error-message is a Truffle and/or Ganache implementation detail, so theoretically, the bug is there and not in the compiler.

However, each one of the following observations leads me to believe that this is in fact a compiler issue:
1. The correct error-message is emitted if we change the input type from a dynamic array `address[] calldata addresses` to a static array `address[2] calldata addresses`.
2. The correct error-message is emitted if we "merge" function `_add` into function `add`, as if some different "stack behavior" took place during runtime, and somehow made this work correctly. I would tend to relate this to the passing by reference of variable `addressSet` from function `add` to function `_add`, but using this variable in function `_add` directly (i.e., not via the reference input variable) results with the same (wrong) error-message.

Please find below the full Truffle test:
```
const MyContract = artifacts.require("MyContract");

const ADDRESS1 = "0x1234567812345678123456781234567812345678";
const ADDRESS2 = "0x8765432187654321876543218765432187654321";

async function assertRevert(promise, message) {
    try {
        await promise;
        throw null;
    }
    catch (error) {
        assert(error, "Expected an error but did not get one");
        assert(error.message.includes(message), `Expected an error with '${message}' but got '${error.message}' instead`);
    }
}

contract("MyContract", (accounts) => {
    let myContract;

    beforeEach(async () => {
        myContract = await MyContract.new();
    });

    it("revert in the 2nd transaction", async () => {
        await myContract.test([ADDRESS1, ADDRESS2]);
        await assertRevert(myContract.test([ADDRESS1, ADDRESS2]), "already exists");
    });

    it("revert in the 1st transaction", async () => {
        await assertRevert(myContract.test([ADDRESS1, ADDRESS1]), "already exists");
    });
});
```
My `truffle-config.js` is the bare minimum; no special configuration other than `optimizer-runs` set to 20000, but all of the above seems to hold also when `optimizer-runs = 200`, and even when `optimizer-enabled = false`, so this is probably not a compiler optimization issue.

**Comment by davidmurdoch** [comment]
@barakman, thanks for opening this issue, looks like an interesting one! We're also leaning towards this being a Solidity compiler bug, and not a Ganache/Truffle bug. We'll keep this issue around on our side so we can follow up on this Soon™😅.

**Comment by haltman-at** [comment]
OK, so I took a look at this and can reproduce the error.  In actual fact it does *not* seem to be a compiler error after all like we thought.  Rather, the reason you're getting an out of gas message, rather than the intended revert message, is because the transaction does in fact run out of gas before it can revert.

(If you're interested, it specifically runs out of gas when attempting to jump back to the test `i < addresses.length` after having incremented `i` from 0 to 1.)

I determined this using Truffle Debugger, but this can also be easily verified by running the transaction with a larger, specified amount of gas; you will indeed get the expected revert message in that case.

So, this isn't a compiler error.  But, that leaves the question of what the actual error here is.  I suppose the error, if there is one, is that Truffle doesn't send enough gas to hit the revert message.  Of course, Truffle determines how much gas to send using `estimateGas`, which I don't believe works for transactions that are bound to fail anyway; I think in such cases Truffle just uses a fixed default which in this case turned out not to be enough.

I'm not certain if I have all of that correct; I believe @fainashalts can say more, I believe it's she who last worked on this and has done a lot more thinking about these gas problems than I have.  So hopefully she can bring some clarity here.  However I should note that I'm not confident there's any good solution to this problem at present; I'm not sure we presently have any good way to determine the amount of gas to use for a transaction that is going to fail anyway, because my understanding is that the underlying tools currently don't presently support that case very well -- which doesn't rule out that possibility that we might be able to handle this anyway, but it does make it harder.

Anyway, thanks for filing this.  I guess we'll have to discuss this.  In the meantime, I suppose you can manually turn up the gas!  Hope that's helpful.

**Comment by barakman** [comment]
@haltman-at : Thank you for your detailed explanation!

From your description, I understand that as part of Truffle implementation, for every transaction, you first call `estimateGas` in order to determine how much gas to pass to the transaction.
And whenever the transaction fails, you have no way of determining how much gas it really needs in order to reach the point of failure, because `estimateGas` itself fails.
So you just pass some arbitrary gas amount, which in this specific case does not allow the transaction to actually reach that point, and instead, the transaction has indeed run out of gas.

What I don't understand, is why you don't simply take the revert reason of your call to `estimateGas` itself.

I actually tried that in my test, and it worked exactly as I'd hoped the original code would:
```
await assertRevert(myContract.test.estimateGas([ADDRESS1, ADDRESS1]), "already exists");
```

So I'm kinda having a problem understanding Truffle's conceptual method of work to begin with.
I'd be happy if you could clarify this for me.

Thanks again :)

**Comment by eggplantzzz** [comment]
Can you clarify a bit more about what you mean by "take the revert reason of your call to `estimateGas` itself? I'm not sure I understand what you are asking.

**Comment by barakman** [comment]
@eggplantzzz : Check out the code example at the bottom of my response (above).

Whenever `myContract.test` is designated to revert, so is `myContract.test.estimateGas`, with the same revert reason.

So instead of executing `myContract.test` with some arbitrary gas amount and hoping that it will suffice for the function to reach the point where it reverts (according to the description given by @haltman-at above), I am asking, why don't you simply execute `myContract.test.estimateGas`?

At least from @haltman-at's description, I am under the impression that prior to executing `myContract.test`, Truffle executes `myContract.test.estimateGas`, in order to get the gas required for executing `myContract.test`.
So my suggestion is, whenever `myContract.test.estimateGas` reverts, just through the revert message right away.

I think it might make more sense to you if you go through the entire thread above (in particularly, though @haltman-at's response).

Thanks :)

**Comment by eggplantzzz** [comment]
Ah I understand, so we'll have to see if we have access to the revert reason when `estimateGas` fails. @haltman-at might know more about whether that is available.

**Comment by eggplantzzz** [comment]
Yep, I just did the same test using a simple contract with a function that fails in the Truffle development console. I think the current approach has been in place for a while, though something did change about the gas estimation maybe 6 months ago (which @fainashalts worked on as @haltman-at may have mentioned above). I suppose we could afford to use the maximum gas if we knew in advance it was a test environment and the user doesn't care about gas...we'll have a discussion and try and see if there is a better solution here. Thanks @barakman!

**Comment by barakman** [comment]
@eggplantzzz:

FYI, this issue significantly hinders the ability to debug contract code using `require` and/or `revert` stubs (added specifically for debug purpose), since these statements are not guaranteed to be reached when the transaction is executed with Truffle.

More specifically, they are guaranteed to be reached only for successful transactions.
But in many (and perhaps even in most) cases, one wishes to debug non-successful transactions.
And there is not much information that one could gather from `exited with an error (status 0) after consuming all gas`...

**Comment by eggplantzzz** [comment]
I hear you @barakman, like I said, we'll have to have a discussion here to hash out details involving a fix etc.

**Comment by minhsieh** [comment]
Hi, I also encountered the same problem today. I used truffle v5.4.26 initially and then upgraded to v5.5.12.
I didn't encounter this problem after upgrading my truffle.

**Comment by haltman-at** [comment]
Hm, so, one approach we could take here is, we could reserve the gas default for `sendTransactionManual`.  After all, if we're not doing a stacktrace, we don't really need to make sure the transaction actually happens, do we?

So, we could have the `send` function throw if `estimateGas` fails in non-debug conditions (and use the error returned from `estimateGas`); but if we're doing `truffle test --stacktrace`, so a debugger is present and we're going to do `sendTransactionManual`, then we can forge ahead and use a default amount of gas.  Since that should never happen in production!

I think that solution should be amenable to everyone. :)

**Comment by haltman-at** [comment]
Oh interesting -- I'm looking through the code right now and it appears that what I just described is in fact how it was *intended* to work, according to the comments.  Just, it doesn't actually at present.  So by fixing this is actually just implementing a bug fix!  Could have saved us some trouble deciding what to do if we'd just read the comments!

**Comment by kevinbluer** [comment]
Hey @barakman, fix for this is now live in [v5.5.18](https://github.com/trufflesuite/truffle/releases/tag/v5.5.18) which has just gone out 👍
